presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,440 @@
+                    .filter(outputSymbols::contains)
+                    .collect(toImmutableList());
+
+            // Cross joins can't filter symbols as part of the join
+            // If we're doing a cross join, use all output symbols from the inputs and add a project node
+            // on top
+            List<Symbol> joinOutputSymbols = sortedOutputSymbols;
+            if (joinFilters.isEmpty()) {
+                joinOutputSymbols = Stream.concat(leftSource.getOutputSymbols().stream(), rightSource.getOutputSymbols().stream())
+                        .collect(toImmutableList());
+            }
+
+            JoinEnumerationResult result =
+                    setJoinNodeProperties(new JoinNode(
+                            idAllocator.getNextId(),
+                            INNER,
+                            leftSource,
+                            rightSource,
+                            ImmutableList.of(),
+                            joinOutputSymbols,
+                            joinFilters.isEmpty() ? Optional.empty() : Optional.of(and(joinFilters)),
+                            Optional.empty(),
+                            Optional.empty(),
+                            Optional.empty()));
+            if (!joinOutputSymbols.equals(sortedOutputSymbols)) {
+                PlanNode resultNode = new ProjectNode(idAllocator.getNextId(), result.planNode.get(), identity(sortedOutputSymbols));
+                result = new JoinEnumerationResult(Optional.of(lookup.getCumulativeCost(resultNode, session, symbolAllocator.getTypes())), Optional.of(resultNode));
+            }
+            return result;
+        }
+
+        /**
+         * This method generates all the ways of dividing  totalNodes into two sets
+         * each containing at least one node. It will generate one set for each
+         * possible partitioning. The other partition is implied in the absent values.
+         * In order not to generate the inverse of any set, we always include the 0th
+         * node in our sets.
+         *
+         * @param totalNodes
+         * @return A set of sets each of which defines a partitioning of totalNodes
+         */
+        @VisibleForTesting
+        static Stream<Set<Integer>> generatePartitions(int totalNodes)
+        {
+            checkArgument(totalNodes >= 2, "totalNodes must be greater than or equal to 2");
+            Set<Integer> numbers = IntStream.range(0, totalNodes)
+                    .boxed()
+                    .collect(toImmutableSet());
+            return Sets.powerSet(numbers).stream()
+                    .filter(subSet -> subSet.contains(0))
+                    .filter(subSet -> subSet.size() < numbers.size());
+        }
+
+        JoinEnumerationResult createJoinAccordingToPartitioning(List<PlanNode> sources, List<Symbol> outputSymbols, Set<Integer> partitioning)
```

Comment:
I prefer outputSymbols.  It represents the required output symbols for the join you're creating.